[JSC] Align stringification algorithm of the Function constructor with the spec

shvaikalesh opened the pull request:
#### ff01344c4f58839cfc9b7381f6416307f9e1e1c5
<pre>
[JSC] Align stringification algorithm of the Function constructor with the spec
<a href="https://bugs.webkit.org/show_bug.cgi?id=190501">https://bugs.webkit.org/show_bug.cgi?id=190501</a>
&lt;rdar://102065151&gt;

Reviewed by Yusuke Suzuki.

This change aligns stringification algorithm of the Function constructor with the spec [1].
In particular, fixes prefixes for (async) generator functions (steps 5 &amp; 7), joins parameters by &quot;,&quot;
instead of &quot;, &quot; (step 10.d.iii), and adds a newline character after the last parameter (step 13).

Since constructFunctionSkippingEvalEnabledCheck() is also being used for creating HTML attribute
event listeners, which uses different stringification algorithm [2], this patch tweaks its signature
to accept the source code instead, making bindings / disabled eval() code a little bit nicer.

Aligns JSC with V8 and SpiderMonkey.

[1]: <a href="https://tc39.es/ecma262/#sec-createdynamicfunction">https://tc39.es/ecma262/#sec-createdynamicfunction</a>
[2]: <a href="https://html.spec.whatwg.org/multipage/webappapis.html#getting-the-current-value-of-the-event-handler">https://html.spec.whatwg.org/multipage/webappapis.html#getting-the-current-value-of-the-event-handler</a> (step 9)

* JSTests/ChakraCore/test/Lib/noargs_2.baseline-jsc:
* JSTests/stress/function-cache-with-parameters-end-position.js:
* JSTests/stress/function-constructor-name.js:
* JSTests/stress/function-constructor-semantics.js:
* JSTests/test262/expectations.yaml: Mark 12 tests as passing.
* LayoutTests/inspector/console/messageAdded-from-named-evaluations-expected.txt:
* LayoutTests/js/dom/function-names-expected.txt:
* LayoutTests/js/dom/script-start-end-locations-expected.txt:
* LayoutTests/js/dom/script-tests/function-names.js:
* Source/JavaScriptCore/API/tests/FunctionOverridesTest.cpp:
* Source/JavaScriptCore/API/tests/testapi.c:
* Source/JavaScriptCore/runtime/FunctionConstructor.cpp:
(JSC::stringifyFunction):
(JSC::constructFunction):
This change is careful to preserve &quot;swallowing&quot; an argument coercion error in case of disabled eval(),
which is intentional since HostEnsureCanCompileStrings happens as early as on step 2 [1].

(JSC::constructFunctionSkippingEvalEnabledCheck):
* Source/JavaScriptCore/runtime/FunctionConstructor.h:
* Source/WebCore/bindings/js/JSLazyEventListener.cpp:
(WebCore::functionParameters):
(WebCore::JSLazyEventListener::JSLazyEventListener):
(WebCore::JSLazyEventListener::initializeJSFunction const):
(WebCore::eventParameterName): Deleted.
* Source/WebCore/bindings/js/JSLazyEventListener.h:

Co-authored-by: Keith Miller &lt;keith_miller@apple.com&gt;
Canonical link: <a href="https://commits.webkit.org/268633@main">https://commits.webkit.org/268633@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3f661ed03698e2738f29ce487220631c2a263f12

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20185 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20619 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21247 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22084 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18836 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/20416 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23862 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/20782 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/20290 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20404 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20307 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17543 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22935 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17476 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18343 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/24611 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/17557 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18553 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18519 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22582 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/19550 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19106 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16222 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/23591 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18310 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/5668 "Found 5 jsc stress test failures: stress/sampling-profiler-microtasks.js.bytecode-cache, stress/sampling-profiler-microtasks.js.dfg-eager, stress/sampling-profiler-microtasks.js.eager-jettison-no-cjit, stress/sampling-profiler-microtasks.js.no-cjit-collect-continuously, wasm.yaml/wasm/stress/simple-inline-stacktrace-with-catch.js.wasm-eager") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22652 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/24848 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2492 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/18931 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5492 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation reviewed (approved):
r=me

MenloDorian commented on Source/JavaScriptCore/runtime/FunctionConstructor.cpp:
```diff
@@ -141,27 +141,27 @@ JSObject* constructFunction(JSGlobalObject* globalObject, const ArgList& args, c
         return nullptr;
     }
-    RELEASE_AND_RETURN(scope, constructFunctionSkippingEvalEnabledCheck(globalObject, args, functionName, sourceOrigin, sourceURL, taintedOrigin, position, -1, functionConstructionMode, newTarget));
+    if (UNLIKELY(code.isNull()))
```
You'll want an `EXCEPTION_ASSERT(!scope.exception());` here to placate the checker because the checker does not know that `globalObject->evalEnabled()` will always return false when an exception occurs and will handle it in the if block above.

MenloDorian replied:
I meant above this line.

MenloDorian replied:
Nevermind, Alexey explained offline that the exception source comes from `stringifyFunction`.  `globalObject->evalEnabled` cannot throw.  So, this is not needed.

However, that EXCEPTION_ASSERT should be:
```
EXCEPTION_ASSERT(!!scope.exception() || !globalObject->evalEnabled());
```
... because that is the condition you checked.

shvaikalesh replied:
Talked to Mark offline, we can clarify & simplify exception checking to `EXCEPTION_ASSERT(!!scope.exception() == code.isNull())`.

Commit:
[JSC] Align stringification algorithm of the Function constructor with the spec
https://bugs.webkit.org/show_bug.cgi?id=190501
<rdar://102065151>

Reviewed by Yusuke Suzuki.

This change aligns stringification algorithm of the Function constructor with the spec [1].
In particular, fixes prefixes for (async) generator functions (steps 5 & 7), joins parameters by ","
instead of ", " (step 10.d.iii), and adds a newline character after the last parameter (step 13).

Since constructFunctionSkippingEvalEnabledCheck() is also being used for creating HTML attribute
event listeners, which uses different stringification algorithm [2], this patch tweaks its signature
to accept the source code instead, making bindings / disabled eval() code a little bit nicer.

Aligns JSC with V8 and SpiderMonkey.

[1]: https://tc39.es/ecma262/#sec-createdynamicfunction
[2]: https://html.spec.whatwg.org/multipage/webappapis.html#getting-the-current-value-of-the-event-handler (step 9)

* JSTests/ChakraCore/test/Lib/noargs_2.baseline-jsc:
* JSTests/stress/function-cache-with-parameters-end-position.js:
* JSTests/stress/function-constructor-name.js:
* JSTests/stress/function-constructor-semantics.js:
* JSTests/test262/expectations.yaml: Mark 12 tests as passing.
* LayoutTests/inspector/console/messageAdded-from-named-evaluations-expected.txt:
* LayoutTests/js/dom/function-names-expected.txt:
* LayoutTests/js/dom/script-start-end-locations-expected.txt:
* LayoutTests/js/dom/script-tests/function-names.js:
* Source/JavaScriptCore/API/tests/FunctionOverridesTest.cpp:
* Source/JavaScriptCore/API/tests/testapi.c:
* Source/JavaScriptCore/runtime/FunctionConstructor.cpp:
(JSC::stringifyFunction):
(JSC::constructFunction):
This change is careful to preserve "swallowing" an argument coercion error in case of disabled eval(),
which is intentional since HostEnsureCanCompileStrings happens as early as on step 2 [1].

(JSC::constructFunctionSkippingEvalEnabledCheck):
* Source/JavaScriptCore/runtime/FunctionConstructor.h:
* Source/WebCore/bindings/js/JSLazyEventListener.cpp:
(WebCore::functionParameters):
(WebCore::JSLazyEventListener::JSLazyEventListener):
(WebCore::JSLazyEventListener::initializeJSFunction const):
(WebCore::eventParameterName): Deleted.
* Source/WebCore/bindings/js/JSLazyEventListener.h:

Co-authored-by: Keith Miller <keith_miller@apple.com>
Canonical link: https://commits.webkit.org/268633@main
